### ISSUE TITLE
First-order finite-diff differentiation with float32 parameters results in floating-point errors

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#1375]
Currently, parameters of type `float32` will incur floating-point errors with finite-diff gradients.  The JAX interface defaults to `float32`, so the issue is most prominent there, but the problem could occur with any interface.

Inside `JacobianTape.numeric_pd`, the relevant parameter does get converted to `np.float64` when [the shift is added](https://github.com/PennyLaneAI/pennylane/blob/23b194ea61dd4158478ce06ef7f4719105f20b57/pennylane/tape/jacobian_tape.py#L265).  But for the calculation of [`"y0"`](https://github.com/PennyLaneAI/pennylane/blob/23b194ea61dd4158478ce06ef7f4719105f20b57/pennylane/tape/jacobian_tape.py#L536), the parameter values are not converted.

Because of this, finite differences will be accurate when `order=2`, but not when `order=1` (the default).

I propose fixing this by casting parameters to `np.float64` for calculation of `"y0"`:

```
                params_f64 = np.array(params, dtype=np.float64)
                options["y0"] = np.asarray(self.execute_device(params_f64, device))
```

This fix assumes every value in `params` is a single scalar. In other words, `params` is not a ragged array. Are we willing to make that assumption?

**Comment by josh146** [comment]
I think your fix makes sense, especially if we are already converting the parameters to `float64`.

> This fix assumes every value in params is a single scalar. In other words, params is not a ragged array. Are we willing to make that assumption?

I am willing to make that assumption, yes. I believe a lot of our abstractions will break if we consider a non-scalar as a 'differentiable parameter'.

I think it should be as simple as editing the failing test in `default.mixed` to ensure that the Hermitian matrix created has `requires_grad=False`.
